### PR TITLE
Remove single quotes in values of Ingestion Server's TSV files

### LINKED ISSUE #3912: Upload Ingestion Server's TSV files to AWS S3
## Description

Corresponds to the [first step](https://docs.openverse.org/projects/proposals/data_normalization/20240227-implementation_plan_catalog_data_cleaning.html#save-tsv-files-of-cleaned-data-to-aws-s3) of the Data Cleaning project. We should have the files for cleaned tags after this is done.

## Additional context

This is part of https://github.com/WordPress/openverse/issues/430.

### PR DESCRIPTION
**Pull request by krysal**
## Fixes

Related to #3912 by @krysal

## Description

This PR fixes several issues on the Ingestion Server and prepares it to upload the files.
- Saves files of cleaned data to a specific temporary directory locally before being send to S3 (this was an @AetherUnbound's previous suggestion)
- Recreates the temp directory before cleaning to avoid mixing values between different data refresh processes
- **Removes single quotes on cleaned values**. Otherwise, the file has URLs surrounded by `'`, which isn't necessary and makes it quite complicated to save the data in a database table later
- Adds stocksnap to TLS_CACHE manually (for some reason, it was being registered as TLS not supported, which is not true)

## Testing Instructions

1. Spin up the services
```
just a && just c
```

2. Make some rows of the image table in the catalog "dirty" by removing the protocol in one or more of the URLs fields (url, creator_url, or foreign_landing_url)

3. Run an image data refresh
```
# Optionally, delete the image index before to avoid the error of "index already exists"
just docker/es/delete-index image-init

just ingestion_server/ingest-upstream "image" "init"
```

4. Check the files are in the container, and check for their content:

```
just exec ingestion_server bash
```

```
ls /tmp/cleaned_data/
cat /tmp/cleaned_data/url.tsv
```

and that the content of each file is the identifier and values without quotes, eg.:

```
0041b4ec-f55e-4cd0-a491-84f41be72232	https://cdn.stocksnap.io/img-thumbs/960w/YUJXHHKSMI.jpg
00252271-e1dc-4faf-b840-7fcc386f2529	https://cdn.stocksnap.io/img-thumbs/960w/7HHEXNL6AQ.jpg
```
While in our bucket, you can find the manually uploaded files containing rows like the following:
```
4e406052-5d90-4c30-87b6-6f1b6ba533e7	'http://musee-mccord.qc.ca/ObjView/MP-1986.85.4.jpg'
3eba813d-6e24-4aa3-9fdb-9242d4bdfc4e	'https://flora-on.pt/Carlina-vulgaris_ori_v6Wn.jpg'
```

## Checklist

- [x] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.
- [N/A] I ran the DAG documentation generator (`just catalog/generate-docs` for catalog PRs) or the media properties generator (`just catalog/generate-docs media-props` for the catalog or `just api/generate-docs` for the API) where applicable.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: ingestion_server/ingestion_server/cleanup.py
Comment (sarayourfriend) [inline ingestion_server/ingestion_server/cleanup.py:312]: TIL the difference between `Path::rmdir` and `shutil.rmtree`. The former can only remove empty directories (the bit I didn't know!).

Comment (krysal) [reply ingestion_server/ingestion_server/cleanup.py]: It's a striking difference but I guess internally it should make sense to keep functions with unique responsibilities.

Comment (sarayourfriend) [reply ingestion_server/ingestion_server/cleanup.py]: For sure!

File: ingestion_server/ingestion_server/cleanup.py
Comment (stacimc) [inline ingestion_server/ingestion_server/cleanup.py:316]: Can you add some comments indicating why these connections are handled differently? I'm not sure I understand the nuance here.

**Review by stacimc — COMMENTED** [review]
I'm able to see the rows I updated getting cleaned, and all the expected logs in the ingestion server locally -- but I don't see the actual file in MinIO 🤔 No errors of any kind. Any ideas what could be going wrong?

File: ingestion_server/ingestion_server/cleanup.py
Comment (krysal) [reply ingestion_server/ingestion_server/cleanup.py:264]: The URL fields had single quotes added twice. As you can see above in the file (lines 117 and 119), I removed the quotes from the cleaned value so they won't appear in the file either (single quotes as the quoting character are problematic for later loading in the DB).

File: ingestion_server/ingestion_server/cleanup.py
Comment (krysal) [reply ingestion_server/ingestion_server/cleanup.py:316]: Sure, I added a comment. Let me know if more is necessary.

In live environments, the connection is allowed via the IAM role policy of the ingestion server instances. You can see that configuration in the following link (only available to maintainers): 

https://github.com/WordPress/openverse-infrastructure/blob/a3c24aefea3c73dcc7b1198a1e3b1c845c4a732b/modules/services/ingestion-server/iam.tf#L57-L78

**Comment by krysal** [comment]
@stacimc That sounds like, potentially, you don't have the `openverse-catalog` bucket that was recently [changed as the default](https://github.com/WordPress/openverse/commit/7c6699d3fa5e2737fa658a94cab79708c2b96a5d) for the catalog. That is not good because it should have raised an error and been logged. I realized my way of checking for the bucket's existence wasn't working. Now I updated and confirmed it checks for the bucket.

You can manually create the bucket through MinIO, or running a `just recreate` should work. Could try again? If the bucket exists and you see no files please share what is in the logs to get any hint.

File: ingestion_server/ingestion_server/cleanup.py
Comment (sarayourfriend) [reply ingestion_server/ingestion_server/cleanup.py:264]: I'm also confused about this :confused: 

Where were the duplicate quotes coming from, exactly? And why does the `clean_value` for tags not need to be quoted but everything else does?

Maybe avoiding the continue would help the clarity of this block:

```py
for field, clean_value in cleaned_data.items():
    if field != "tags":
        # Save cleaned values for later
        # (except for tags, which take up too much space)
        cleaned_values[field].append((identifier, clean_value))

    update_field_expressions.append(f"{field} = {clean_value}")
```

Never mind that I don't understand where the difference is coming from for the format of the string added to `update_field_expressions`, this version is a lot easier to understand, to me.

File: ingestion_server/ingestion_server/cleanup.py
Comment (sarayourfriend) [inline ingestion_server/ingestion_server/cleanup.py:295]: Kind of just a nit, but I usually see usage of the `/` operator to join path segments, rather than `joinpath`, although we seem to have a mix of both in our project (maybe more `/` versions, though?). Anyway, just a nit/nothing.

```suggestion
        with open(TMP_DIR / f"{field}.tsv", "a", encoding="utf-8") as f:
```

File: ingestion_server/ingestion_server/cleanup.py
Comment (sarayourfriend) [inline ingestion_server/ingestion_server/cleanup.py:309]: Suggestion to clarify this comment. I wasn't sure what "it" referred to (might have both been MinIO and boto3?), and the reason for the difference between local and live wasn't clear (like, why MinIO matters).

```suggestion
    """
    Retrieve a correctly configured boto3 S3 resource.
    
    Locally, we use MinIO to emulate S3, so we must specify the endpoint and credentials.
    On live environments, S3 itself is used, and authentication works via the instance profile.
    """
```

Although, overall, I do think this would be clearer if we didn't use default values here, and instead required setting them, and allowed them to be `None` when not configured in the environment. Then the code wouldn't need to worry about the implementation detail _at all_. I'll share a suggestion for that code in my next comment.

File: ingestion_server/ingestion_server/cleanup.py
Comment (sarayourfriend) [inline ingestion_server/ingestion_server/cleanup.py:320]:
```suggestion
    if config("ENVIRONMENT", default="local") == "local":
        return boto3.resource(
            "s3",
            endpoint_url=config("AWS_S3_ENDPOINT", default="http://s3:5000"),
            aws_access_key_id=config("AWS_ACCESS_KEY_ID", default="test_key"),
            aws_secret_access_key=config(
                "AWS_SECRET_ACCESS_KEY", default="test_secret"
            ),
        )

    return boto3.resource(
        "s3",
        endpoint_url=config("AWS_S3_ENDPOINT", default=None),
        aws_access_key_id=config("AWS_ACCESS_KEY_ID", default=None),
        aws_secret_access_key=config("AWS_SECRET_ACCESS_KEY", default=None),
        region_name=config("AWS_REGION", default=None),
    )
```

All these kwargs are defined with None as the default in the implementation, so we can safely explicitly pass null (if these variables aren't defined in the environment), and they'll just be ignored.

https://boto3.amazonaws.com/v1/documentation/api/latest/_modules/boto3/session.html#Session.resource

That way the code doesn't have to worry about implementation details of local/live environments, and we get rid of this confusion altogether...

File: ingestion_server/ingestion_server/cleanup.py
Comment (sarayourfriend) [inline ingestion_server/ingestion_server/cleanup.py:339]: In what case is this file missing not an error condition (in other words, why can we just `continue` here rather than raise an exception)? The intention for this loop would be a lot safer and more explicit if instead of looping through `fields`, we only looped through the specific ones we want to upload.

**Review by sarayourfriend — CHANGES_REQUESTED** [review]
Requesting changes to simplify the `get_s3_resource` function (or even just remove it), clarify further the changes regarding `update_field_expressions`, and clarify the loop in `_upload_to_s3` (specifically the case where the file does not exist).

File: ingestion_server/ingestion_server/cleanup.py
Comment (krysal) [reply ingestion_server/ingestion_server/cleanup.py:295]: I took the idea from your recent changes in managing the files for waveform. I like how explicit it is in the `with` line. 

https://github.com/WordPress/openverse/blob/d562f4719aefdf7d90c13ca6738a5e8c27ce352f/api/api/utils/waveform.py#L79

I found the form with ` / ` handy for longer paths in a separate line.

File: ingestion_server/ingestion_server/cleanup.py
Comment (krysal) [reply ingestion_server/ingestion_server/cleanup.py:339]: Once we start fixing the rows, we will reach the point where there is nothing to clean, so there is no file to upload in that case, and the `tags` field won't have a file here either.

File: ingestion_server/ingestion_server/cleanup.py
Comment (krysal) [reply ingestion_server/ingestion_server/cleanup.py:320]: That's a good observation! Thank you. I agree that we can simplify this a lot, though it'd require setting these environment variables locally for the ingestion server to work.

File: ingestion_server/ingestion_server/cleanup.py
Comment (krysal) [reply ingestion_server/ingestion_server/cleanup.py:262]: > Where were the duplicate quotes coming from, exactly?

The cleaned values have quotes added in the return of the cleaning function:
https://github.com/WordPress/openverse/blob/d562f4719aefdf7d90c13ca6738a5e8c27ce352f/ingestion_server/ingestion_server/cleanup.py#L108-L111

>  And why does the clean_value for tags not need to be quoted but everything else does?

The `psycopg2.extras.Json` function for tags already adapts a Python object to :sql:`json` data type, so they don't need extra quotes wrapping.

The difference in the string added to `update_field_expressions` is the quotes. I assume this was previously done that way to avoid this confusion, but it's necessary to have the cleaned values without these quotes in the file so they don't interfere with the copy upload to a DB table later.

An alternative would be to perform more string manipulation to remove the quotes before saving the rows in files, but this seemed unnecessary and more prone to error to me.

File: ingestion_server/ingestion_server/cleanup.py
Comment (krysal) [reply ingestion_server/ingestion_server/cleanup.py]: I ended up removing the function per Sara's suggestion so it should be simpler now.

File: ingestion_server/ingestion_server/cleanup.py
Comment (krysal) [reply ingestion_server/ingestion_server/cleanup.py]: Minio is what we use to emulate and test some AWS operations. In this case, it's the replacement for S3. Otherwise, we would need to use some test live resources, which would complicate things a lot.

File: ingestion_server/ingestion_server/cleanup.py
Comment (krysal) [reply ingestion_server/ingestion_server/cleanup.py]: Moved to #4529.

**Comment by krysal** [comment]
@stacimc @sarayourfriend I split the PR into two parts since many things were happening here and it's easier to explain by simplifying it. Here, I left the fix for the quotes in the TSV files, updated the testing instructions, and opened #4529 for the upload to S3. I hope I have answered all your questions. Let me know if anything needs further clarification.

**Review by stacimc — CHANGES_REQUESTED** [review]
@krysal I'm not sure what's happening, but I'm getting really strange behavior when testing this. Locally in my catalog I updated three different records, removing the protocol from the `url` on one, the `creator_url` on the second, and the `foreign_landing_url` on the third. When I ran the data refresh only the `foreign_landing_url` actually got cleaned, ~and I still am not seeing the file in minio even though I see no errors in my ingestion server logs~. Edit: I see that you moved the s3 uploading out of this PR so that part makes sense.

I didn't have much time to look further into this, but is anyone else able to replicate my results?

File: ingestion_server/ingestion_server/cleanup.py
Comment (sarayourfriend) [reply ingestion_server/ingestion_server/cleanup.py:262]: I agree with Staci, but I also don't want to block on it because of this code being deleted soon anyway :shrug:

**Review by sarayourfriend — APPROVED** [review]
I wasn't able to replicate Staci's results. I modified an image in the catalog to remove the scheme from all three URLs, and all three have separate files correctly adding the scheme.

I think there are ways to make the intention of the code clearer, but as I said, I am unwilling to block the PR given the code will be gone soon anyway. That said, I would be more firm in requesting clarification comments or reorganisation of that loop if this code was long term, so if similar code will be introduced elsewhere, I would block on its clarification. Just wanted to clarify it's a matter of the circumstance, not that I would encourage a pattern of ignoring clarity in reviews.

File: ingestion_server/ingestion_server/cleanup.py
Comment (krysal) [reply ingestion_server/ingestion_server/cleanup.py:266]: I updated the comment. If it's still not clear, it's open to suggestions. I didn't think it would bring so much confusion 😅

**Comment by krysal** [comment]
@stacimc Did you try doing a full `just recreate` before all the steps? ... @dhruvkb are you able to test and review the PR?

---

@sarayourfriend unrelated but I was about to suggest using `./ov just recreate` but it didn't work for me. I got the following:
```
➜ ./ov just recreate
Unable to find image 'openverse-dev_env:latest' locally
docker: Error response from daemon: pull access denied for openverse-dev_env, repository does not exist or may require 'docker login'.
See 'docker run --help'.
```
The plain `just recreate` worked.

**Review by dhruvkb — APPROVED** [review]
The code changes make sense based on the PR description. I followed the testing instructions and verified the correct output in `/tmp`.

**Comment by sarayourfriend** [comment]
@krysal that's because of @dhruvkb's PR to https://github.com/WordPress/openverse/pull/4526.

**Comment by stacimc** [comment]
@krysal I did recreate, but if three other people are unable to reproduce I'm willing to believe something was going awry in my environment. No objections from me 👍

**Pull request merged**
